### PR TITLE
feat(observability): tag proxy spans with model and upstream status (#281)

### LINKED ISSUE #281: feat(observability): tag proxy spans with model and upstream status, capture upstream errors in Sentry
## Problem / motivation

A 2026-07-30 production incident (auto-mode classifier requests to `claude-opus-5` reported "temporarily unavailable" through the gateway) could not be root-caused, because two independent gaps closed off every diagnostic path:

1. **Deployment gap (context only, not a code fix):** the deployment logs only to a foreground tty — nothing is persisted, so there was no log to grep after the fact.
2. **Instrumentation gap (the actual bug here):** shunt's Sentry `proxy_request` span records only `http.request.method`, `path`, and duration. It carries **no requested model id, no upstream/resolved provider, and no upstream HTTP response status**. Upstream 4xx/5xx responses are mapped straight to shaped JSON error responses (per `AGENTS.md`'s Anthropic-error-shape convention) without ever going through Sentry's error-capture path.

Round-5 investigation evidence (read-only Sentry API, org `passion-factory`, project `shunt`, release `shunt-gateway@0.28.0`):
- 412 `proxy_request` spans existed in the ~3h incident window; **every span's `span.status` was the constant `"unknown"`** (never `ok`/`error`).
- Tried 9 candidate span/tag field names — `gen_ai.request.model`, `model`, `model_id`, `request.model`, `http.status_code`, `status_code`, `span.status`, `trace.status`, `upstream.status`, `http.response.status_code` — **all resolved to `null`** across every span in the window.
- 0 error-level Sentry events existed anywhere in the project for that window (errors dataset `count()` = 0), and searching all-time issues for `claude-opus-5`, `claude-sonnet-5`, `overloaded`, `rate_limit`, `not_found_error` returned 0 matches — only 3 unrelated infra issues exist in the whole project (bind-failure, config-reload-failure x2).

Net effect: even with `traces_sample_rate = 1.0` capturing every request, Sentry cannot answer "which model was requested" or "what did the upstream return" for any single request — the fields simply don't exist. A repeat of this incident today would be just as unroot-causeable from Sentry as this one was.

Related/prior art (not duplicates, but adjacent — worth cross-referencing during implementation):
- #65 confirms `proxy_request` is presently always a root span (no inbound traceparent extraction yet) — relevant context for where these attributes would be recorded.
- #189 (inbound OTLP ingest, part of epic #186) is about *receiving* telemetry from clients, not shunt's own outbound span attributes — different problem, same general observability epic.

## Proposal

In `src/proxy.rs`, around the existing `proxy_request` span (`code.file.path: src/proxy.rs`, `code.line.number: 49` as of `shunt-gateway@0.28.0`):

1. Record the requested model id and the resolved upstream/provider as span attributes. Use the OTel GenAI semantic convention name `gen_ai.request.model` as the attribute for the model id (matches the field this investigation already probed for and matches prevailing OTel practice), plus a `shunt.provider` (or similar) attribute for the resolved upstream.
2. Record the upstream response status code on the span, and set the span's own `status` (`ok`/`error`) accordingly instead of the current constant `"unknown"`.
3. Emit a Sentry error/warning-level event when the upstream returns 5xx, and on 429/529 quota-shed responses, tagged with model + status + provider — but **never** message content, request/response headers, or credentials.
4. Consider a config flag if there's a cardinality or PII concern with always tagging model id (e.g. gate behind an opt-in `[server.sentry]`-style flag, consistent with this repo's existing pattern of opt-in observability surfaces).

## Acceptance criteria

- A failed upstream request (4xx/5xx/429/529) is findable in Sentry by model id + status code.
- A repeat of the 2026-07-30 incident would be root-causeable from Sentry alone, without needing server-side log access.
- No message content, headers, or credentials are ever sent to Sentry.

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Closes #281. Goal: make an upstream failure (5xx, or 429/529 quota/overload) root-causeable from Sentry alone — which model, which provider, what status — without cross-referencing local logs.

Two independent mechanisms, since they reach Sentry through different paths (see the module doc on `src/observability.rs` for the full rationale):

- **Span tagging** (`record_requested_model`, `record_span_outcome` in `src/observability.rs`): fills in `tracing::field::Empty` fields declared on the `proxy_request` (`src/proxy.rs`) and `codex_endpoint_request` (`src/codex_endpoint.rs`, the sibling inbound-Codex path — confirmed it shares the same span+instrument pattern) spans at creation time: `gen_ai.request.model` (OTel GenAI semconv), `shunt.provider`, `http.response.status_code`, and `otel.status_code` (`error` on 4xx/5xx). Sentry's own span/transaction *status* has no tracing-field convention (unlike `otel.status_code`), so it's set directly via `sentry::configure_scope(..).get_span().set_status(..)` — a no-op whenever no Sentry span is active.
- **Event capture** (`capture_upstream_outcome`): a Sentry error/warning event for 5xx and 429/529 responses, built with `sentry::capture_message` directly rather than `tracing::error!`/`warn!`. The Sentry tracing layer in `main.rs` has no custom `EventFilter`, so its default demotes `WARN` to a breadcrumb, not an event — routing through the macros would silently drop the 429/529 signal. Calling the Sentry API directly avoids depending on that filter and keeps this change out of `main.rs`'s subscriber wiring.

Both mechanisms carry only the requested model id, resolved provider, and numeric upstream status — never request/response bodies, headers, or credentials.

Status is read at response-header time in both `src/proxy/failover.rs::forward` (all return paths, plus `count_tokens_response`) and `src/codex_endpoint.rs::forward`; no SSE buffering was introduced.

**Config**: no new public config key was added. A flag to gate this tagging was considered and deferred — the existing `[sentry] traces_sample_rate` / `[otel] traces` opt-ins already gate span export, and event capture only ever emits a handful of low-cardinality tags (no new privacy surface), so a second flag would add configuration surface without a corresponding new risk to guard.

**Docs**: updated `site/src/content/docs/guides/opentelemetry.mdx`'s trace-signal row to describe the new span attributes (the only affected surface — no config/endpoint/CLI/provider changed, so README/other docs are unaffected). Other locale files (`ko`/`ja`/`zh-cn`) fall back to the English page automatically per the site's i18n routing, so no translation is required for this to render correctly.

## Test plan

- [x] `cargo build --all-features`
- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --all-targets --all-features -- -D warnings` — zero warnings
- [x] `cargo test --all-features --workspace` — 1040 passed, 0 failed, 1 ignored
- New coverage in `src/observability.rs`: `should_capture_upstream_status_covers_all_bands` (429/529 → Warning, 5xx → Error, else None — including the 529-is-also-`is_server_error()` edge case), `sentry_span_status_maps_status_bands`, and two span-field-recording tests using a custom `tracing_subscriber::Layer` that captures `on_record` values directly (asserts `gen_ai.request.model`, `shunt.provider`, `http.response.status_code`, `otel.status_code` all land on the current span).

Draft PR — AI reviewers will run once this is marked ready for review.

---
## Summary by cubic
Tags `proxy_request` and `codex_endpoint_request` spans with the requested model, provider, and upstream HTTP status, and sends Sentry events for 5xx and 429/529 so upstream failures are diagnosable without local logs (#281). Also sanitizes the client-supplied model tag with bounded work (strip control chars, 128-char cap).

- **New Features**
  - Span tags: `gen_ai.request.model`, `shunt.provider`, `http.response.status_code`, `otel.status_code` (`error` on 5xx). Set when response headers are known in both proxy and Codex paths; Sentry span status is set directly.
  - Events: 5xx → error, 429/529 → warning. Sent via `sentry::capture_message`; tagged with `model`/`provider`/`upstream_status` only.
  - Docs/Tests: updated OTEL guide and config docs (all locales), running guide/`SentryConfig`, and `shunt.toml.example`. Added unit tests for event/span mapping and a live failover test using an isolated Sentry Hub; enabled `sentry`’s `test` feature.

- **Bug Fixes**
  - OTel: only set `otel.status_code=error` on 5xx; 4xx leaves it `ok`.
  - Safety: sanitize `model` before tagging (strip control chars, truncate to 128 chars with short-circuiting to bound allocations, fallback to `"invalid"`).
  - Performance: bound control-char detection to the first 128 chars to avoid scanning large inputs.

<sup>Written for commit d8fdf1febe022b45dfc3eeded027a5e02d29914f. Summary will update on new commits.</sup>

### CODE REVIEW

File: src/observability.rs [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    let span = tracing::Span::current();
+    span.record("shunt.provider", provider);
+    span.record("http.response.status_code", status.as_u16());
+    span.record(
+        "otel.status_code",
+        if status.is_client_error() || status.is_server_error() {
+            "error"
+        } else {
+            "ok"
+        },
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/observability.rs]: Good catch — otel.status_code=error for 4xx violates OTel semconv for HTTP server spans (only 5xx/transport failures should mark a span error). Applied exactly as suggested in record_span_outcome (src/observability.rs), added a regression test for the 4xx→ok case, and left sentry_span_status untouched since Sentry's own SpanStatus convention already distinguishes 4xx bands independently of otel.status_code. Pushed in 20affcf.

File: site/src/content/docs/guides/opentelemetry.mdx [hunk 2]
```diff
@@ -37,7 +37,7 @@ Setting `endpoint = ""` (e.g. `SHUNT_OTEL__ENDPOINT=""`) disables export again w
 
 | Signal | What's exported | Notes |
 | :-- | :-- | :-- |
-| **Traces** | The per-request `proxy_request` span | Head-based sampling via `sample_ratio`. Low-cardinality; no request/response bodies. |
+| **Traces** | The per-request `proxy_request` span (and `codex_endpoint_request` on the inbound Codex endpoint), tagged with `gen_ai.request.model`, `shunt.provider`, `http.response.status_code`, and `otel.status_code` (`error` on 4xx/5xx) | Head-based sampling via `sample_ratio`. Low-cardinality; no request/response bodies, headers, or credentials. |
```
[the remark this answers is not in the document]
Comment (amondnet) [reply site/src/content/docs/guides/opentelemetry.mdx]: Applied — updated the traces-table row to '(error on 5xx)' to match the record_span_outcome fix. Checked the other doc surfaces that could carry this same claim: the locale opentelemetry.mdx copies (ko/ja/zh-cn) don't have this traces-table row yet (not translated), and configuration.md across all locales only describes the separate Sentry-event mechanism (capture_upstream_outcome, 5xx/429/529), not otel.status_code, so nothing else needed updating. Pushed in 20affcf.

File: src/observability.rs [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+/// no client configured. Uses `sentry::capture_message` directly rather than a
+/// `tracing::error!`/`warn!` macro so this does not depend on overriding the
+/// `sentry` tracing layer's default `EventFilter` (which otherwise downgrades
+/// `WARN` to a breadcrumb, not an event) — see the module docs.
+pub(crate) fn capture_upstream_outcome(provider: &str, model: &str, status: StatusCode) {
+    let Some(level) = should_capture_upstream_status(status) else {
+        return;
+    };
+    let message = match level {
+        Level::Error => "upstream returned an error response",
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/observability.rs:216]: Confirmed — the inbound Codex endpoint's model field is unvalidated free text (ModelView, request body forwards upstream verbatim) and it was reaching both the Sentry event tag (capture_upstream_outcome) and the span field (record_requested_model) unbounded. Added sanitize_model_tag in src/observability.rs: strips control characters, truncates to 128 chars (well under Sentry's 200-char tag limit), falls back to 'invalid' when the sanitized value is empty. Wired into both call sites so proxy.rs and codex_endpoint.rs get the bound for free. Covered with boundary unit tests (exactly-at-limit, over-limit, control chars, empty/control-only) plus an end-to-end test proving an oversized/control-char model is actually sanitized before it lands in a captured Sentry event's tag. Pushed in 1277e51.

File: src/observability.rs [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+    sentry::with_scope(
+        |scope| {
+            scope.set_tag("model", model.as_ref());
+            scope.set_tag("provider", provider);
+            scope.set_tag("upstream_status", status_code);
+        },
+        || {
+            sentry::capture_message(message, level);
+        },
+    );
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/observability.rs:230]: This is a false positive — the two-closure call is the correct, documented sentry-core API, not a bug.

Evidence:
1. sentry-core 0.48.4 (the version pinned in Cargo.lock — verified via `grep -A2 '^name = "sentry-core"$' Cargo.lock`), src/api.rs:184-188:
   `pub fn with_scope<C, F, R>(scope_config: C, callback: F) -> R where C: FnOnce(&mut Scope), F: FnOnce() -> R`
   — it takes exactly two closures by design: one to configure a temporarily-pushed scope, one to run inside it.
2. The crate's own doctest (same file, lines 174-179) uses the identical pattern: `sentry::with_scope(|scope| scope.set_level(...), || sentry::capture_message(...))`.
3. Empirically: this exact code built and passed `cargo clippy --all-targets --all-features -- -D warnings` and the full workspace test suite locally, and CI's 'fmt · clippy · test' check is green on commit 1277e51 (https://github.com/pleaseai/shunt/actions — job 90854751617).

No code change made — resolving as a false positive.

File: src/observability.rs [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+        Cow::Owned(cleaned.chars().take(MAX_MODEL_TAG_LEN).collect())
+    } else {
+        cleaned
+    };
+    if truncated.is_empty() {
+        Cow::Borrowed("invalid")
+    } else {
+        truncated
+    }
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/observability.rs:110]: Agreed, applied — this closes a real gap: the previous version always re-collected into a fresh String to truncate, even when the value was already Cow::Owned from the control-strip or short enough to slice in place, so a client sending an oversized-but-otherwise-clean model string (the primary case this function guards against) paid a second allocation for nothing. sanitize_model_tag now finds the truncation byte boundary via char_indices and either slices a Cow::Borrowed or String::truncates a Cow::Owned in place, so at most one allocation is ever paid (the one the control-character strip already needed). Existing boundary tests (exactly-at-limit, over-limit, control chars, empty) pass unchanged since the observable behavior is identical. Pushed in 54b61fc.

File: src/observability.rs [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+                Cow::Owned(s)
+            }
+        };
+    }
+    if cleaned.is_empty() {
+        Cow::Borrowed("invalid")
+    } else {
+        cleaned
+    }
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/observability.rs:110]: Reviewed and rejecting this for now — not a gap in the current fix, but a deliberate scope boundary.

1. The diagnostic purpose of this PR (issue #281) is precisely to surface the literal requested model id in Sentry when an upstream failure occurs — the motivating incident (an auto-mode classifier failing on a specific model id) was only root-causeable because the exact string mattered (alias vs. bare id). Hashing or allowlisting the tag would destroy that diagnostic value.
2. The exposure surface is already bounded by this PR: the value is control-character-stripped and length-capped (128 chars, see MAX_MODEL_TAG_LEN), it's only tagged on upstream-failure events/spans (not every request), and it flows only to the operator's own Sentry project (self-hosted DSN, no third party beyond the operator's own choice) — the same trust domain as the gateway's own logs.
3. A stricter policy (allowlist/hash, likely config-gated so operators can opt into a hardened mode) is a deliberate product decision and out of scope for this PR. It can be proposed as a follow-up; #287 already tracks adjacent observability follow-up scope.

No code change. Resolving.

File: src/observability.rs [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+/// and inflating tag cardinality/size. Falls back to `"invalid"` when the
+/// sanitized value is empty (missing/blank model, or a value that is nothing
+/// but control characters). Pure so it is unit-testable in isolation.
+fn sanitize_model_tag(model: &str) -> Cow<'_, str> {
+    let needs_stripping = model.chars().any(char::is_control);
+    let mut cleaned: Cow<'_, str> = if needs_stripping {
+        Cow::Owned(model.chars().filter(|c| !c.is_control()).collect())
+    } else {
+        Cow::Borrowed(model)
+    };
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/observability.rs]: Confirmed and applied — good catch. The strip-then-collect path did allocate proportional to the full client-supplied input before truncation ever ran, so a control character anywhere in a multi-KB/MB model string (up to the request-body limit) forced an unbounded allocation.

Fixed in 67b8d0d using the same idea you suggested — `.take(MAX_MODEL_TAG_LEN)` — applied during the `filter()` collection so `chars()` short-circuits once enough non-control characters have been produced:

```rust
Cow::Owned(
    model
        .chars()
        .filter(|c| !c.is_control())
        .take(MAX_MODEL_TAG_LEN)
        .collect(),
)
```

The no-control-chars/truncate-only branch was folded into the same expression (dropped the redundant intermediate `cleaned` rebind from the prior commit) since `char_indices().nth(MAX_MODEL_TAG_LEN)` on the original slice is equally bounded (also short-circuits). Added a boundary test with a 64 KiB pathological input asserting the output is still capped at MAX_MODEL_TAG_LEN; all existing sanitize_model_tag tests pass unchanged. Gates green: fmt, clippy -D warnings, full workspace test (1053 passed/0 failed/1 ignored).

File: src/observability.rs [hunk 8]
```diff
@@ -0,0 +1,604 @@
...
+        Cow::Borrowed(&model[..byte_idx])
+    } else {
+        Cow::Borrowed(model)
+    };
+    if cleaned.is_empty() {
+        Cow::Borrowed("invalid")
+    } else {
+        cleaned
+    }
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/observability.rs:110]: Agreed and applied, though with a smaller diff than the suggestion: only the first `MAX_MODEL_TAG_LEN` raw characters can ever affect the output. If none of them is a control character, the strip-and-collect branch below would keep exactly that prefix unfiltered anyway (`take(MAX_MODEL_TAG_LEN)` stops right there, since all of them pass the filter) — identical to the plain-truncate branch. So a control character beyond that window can never change the result, and the detection scan can safely stop there too:

```rust
let needs_stripping = model.chars().take(MAX_MODEL_TAG_LEN).any(char::is_control);
```

This bounds the common-case (no control chars) scan to O(MAX_MODEL_TAG_LEN) rather than O(input length), without needing to fold truncation-point tracking into the same loop — the existing char_indices().nth(MAX_MODEL_TAG_LEN) truncation path is already bounded on its own. Pushed in d8fdf1f, with a new test (sanitize_model_tag_ignores_a_control_character_beyond_the_truncation_window) proving a late control character doesn't change the output. Gates green: fmt, clippy -D warnings, full workspace test (1054 passed/0 failed/1 ignored).
